### PR TITLE
CI for Configuration and Building (Windows)

### PR DESCRIPTION
**Pull request by NicerNewerCar**
Addresses #65

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# This pipe line will be triggered when a pull request is created and will validiate that the CUDA and OpenCL versions of Autoscoper can configure and build successfully.
+
+name: CI - Ubuntu
```
Comment (jcfr) [inline .github/workflows/ci.yml:3]:
```suggestion
name: CI
```

Comment (jcfr) [reply .github/workflows/ci.yml:3]: System being used for testing will be part of the job name

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      uses: jurplel/install-qt-action@v3
+      with:
+        version:      ${{ env.QT_VERSION }}
+        host:         linux
+        target:       desktop
+        arch:         gcc_64
+        dir:          /opt/qt
+        modules:      webassembly qtwebengine sources qtwebglplugin qtscript debug_info
+        setup-python: false
+    - name: Install CUDA
```
Comment (jcfr) [inline .github/workflows/ci.yml:30]:
```suggestion
    - name: Install CUDA
      if: ${{ matrix.rendering_backend == 'CUDA' }}
      uses: Jimver/cuda-toolkit@v0.2.8
```

This assumes the matrix will be defined using something like this


```yaml
jobs:
  autoscoper:
    strategy:
      matrix:
        rendering_backend: ['CUDA', 'OpenCL']
    steps:
      ....
```

**Review by jcfr — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    - name: Install Qt
+      uses: jurplel/install-qt-action@v3
+      with:
+        version:      ${{ env.QT_VERSION }}
+        host:         linux
+        target:       desktop
+        arch:         gcc_64
+        dir:          /opt/qt
+        modules:      webassembly qtwebengine sources qtwebglplugin qtscript debug_info
+        setup-python: false
```
Comment (jcfr) [inline .github/workflows/ci.yml:34]:
```suggestion
        version: ${{ env.QT_VERSION }}
        host: linux
        target: desktop
        arch: gcc_64
        dir: /opt/qt
        setup-python: false
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      uses: jurplel/install-qt-action@v3
+      with:
+        version: ${{ env.QT_VERSION }}
+        host: linux
+        target: desktop
+        arch: gcc_64
+        dir: /opt/qt
+        setup-python: false
+    - name: Install Backend # cuda-tk provides the CUDA and OpenCL backends for Autoscoper
+      uses: Jimver/cuda-toolkit@v0.2.8
```
Comment (jcfr) [inline .github/workflows/ci.yml:35]:
```suggestion
      uses: Jimver/cuda-toolkit@v0.2.8
      id: cuda-toolkit
      with:
        cuda: '10.1'
```

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    - name: Configure CMake
+      # Configure CMake in a 'build' subdirectory. `CMAKE_BUILD_TYPE` is only required if you are using a single-configuration generator such as make.
+      # See https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html?highlight=cmake_build_type
+      run: cmake -S ${{github.workspace}} -B ${{github.workspace}}/build -DCMAKE_BUILD_TYPE=${{env.BUILD_TYPE}} -DAutoscoper_RENDERING_BACKEND=${{matrix.rendering_backend}} -DAutoscoper_SUPERBUILD=ON
```
Comment (jcfr) [inline .github/workflows/ci.yml:39]:
```suggestion
      run: |
        cmake \
          -S ${{github.workspace}} \
          -B ${{github.workspace}}/build \
          -DCMAKE_BUILD_TYPE=${{env.BUILD_TYPE}} \
          -DAutoscoper_RENDERING_BACKEND=${{matrix.rendering_backend}} \
          -DAutoscoper_SUPERBUILD=ON
```

See https://stackoverflow.com/questions/59954185/github-action-split-long-command-into-multiple-lines

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    - name: Build 
+      # Build your program with the given configuration
+      run: cmake --build ${{github.workspace}}/build --config ${{env.BUILD_TYPE}}
```
Comment (jcfr) [inline .github/workflows/ci.yml:42]:
```suggestion
      run: |
        cmake \
          --build ${{github.workspace}}/build \
          --config ${{env.BUILD_TYPE}}
```

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+name: CI
+
+on:
+  pull_request:
+  push:
+    branches:
+    - main
+
+
```
Comment (jcfr) [inline .github/workflows/ci.yml:11]:
```suggestion
```

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -0,0 +1,42 @@
+# This pipe line will be triggered when a pull request is created and will validiate that the CUDA and OpenCL versions of Autoscoper can configure and build successfully.
```
Comment (jcfr) [inline .github/workflows/ci.yml:1]:
```suggestion
```

**Comment by NicerNewerCar** [comment]
@jcfr Got the build working for windows, see [CI #16](https://github.com/BrownBiomechanics/Autoscoper/actions/runs/3742604265), the build on ubuntu is failing it looks like the version of opengl that github has on there runners does not work with glew see [CI #13](https://github.com/BrownBiomechanics/Autoscoper/actions/runs/3741892663/jobs/6352106123). 

We also cannot run ubuntu and windows from the same script because there are a couple differences between the two( ` to split line on windows vs \ to split line on ubuntu, and using a difference compiler on the two OSs)

Note: This pipeline takes > 1 hour to run, there are 2,000 minutes/month included for free with github and Windows runners consume 2x the minutes of linux runners (ie. 10 minutes on a windows runner is equivalent to 20 minutes on a linux runner).

See [About billing for GitHub Actions](https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions).

**Comment by NicerNewerCar** [comment]
Closing this since we are using [SA^M](https://github.com/BrownBiomechanics/SlicerAutoscoperM) to perform CI

**Pull request closed without merging**